src/dtls.c

```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
+        suites.hashSigAlgoSz = (word16)sigAlgs.size;
+        XMEMCPY(suites.hashSigAlgo, sigAlgs.elements, sigAlgs.size);
+        haveSA = 1;
+    }
+
+    /* Supported groups */
+    ret = TlsxFindByType(&tlsx, TLSX_SUPPORTED_GROUPS,
+                         ch->extension, &tlsxFound);
+    if (ret != 0)
+        goto dtls13_cleanup;
+    if (tlsxFound) {
+        ret = TLSX_SupportedCurve_Parse(ssl, tlsx.elements,
+                                     (word16)tlsx.size, 1, &parsedExts);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        haveSG = 1;
+    }
+
+    /* Key share */
+    ret = TlsxFindByType(&tlsx, TLSX_KEY_SHARE,
+                         ch->extension, &tlsxFound);
+    if (ret != 0)
+        goto dtls13_cleanup;
+    if (tlsxFound) {
+        ret = TLSX_KeyShare_Parse_ClientHello(ssl, tlsx.elements,
+                                        (word16)tlsx.size, &parsedExts);
+        if (ret != 0)
+            goto dtls13_cleanup;
+        haveKS = 1;
+    }
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+    /* Pre-shared key */
+    ret = TlsxFindByType(&tlsx, TLSX_PRE_SHARED_KEY, ch->extension, &tlsxFound);
+    if (ret != 0)
+        goto dtls13_cleanup;
+    if (tlsxFound) {
+        /* Let's just assume that the binders are correct here. We will
+         * actually verify this in the stateful part of the processing
+         * and if they don't match we will error out there anyway. */
+        byte modes;
+
+        /* Ask the user for the ciphersuite matching this identity */
+        if (TLSX_PreSharedKey_Parse_ClientHello(&parsedExts,
+                tlsx.elements, tlsx.size, ssl->heap) == 0)
+            FindPskSuiteFromExt(ssl, parsedExts, &pskInfo, &suites);
+        /* Revert to full handshake if PSK parsing failed */
+
+        if (pskInfo.isValid) {
+            ret = TlsxFindByType(&tlsx, TLSX_PSK_KEY_EXCHANGE_MODES,
+                                 ch->extension, &tlsxFound);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if (!tlsxFound)
+                ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+            ret = TLSX_PskKeyModes_Parse_Modes(tlsx.elements, tlsx.size,
+                                              client_hello, &modes);
+            if (ret != 0)
+                goto dtls13_cleanup;
+            if ((modes & (1 << PSK_DHE_KE)) &&
+                    !ssl->options.noPskDheKe) {
+                if (!haveKS)
+                    ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+                doKE = 1;
+            }
+            else if ((modes & (1 << PSK_KE)) == 0) {
+                    ERROR_OUT(PSK_KEY_ERROR, dtls13_cleanup);
+            }
+            usePSK = 1;
+        }
+    }
+#endif
+
+#if defined(HAVE_SESSION_TICKET) || !defined(NO_PSK)
+    if (usePSK && pskInfo.isValid) {
+        cs.cipherSuite0 = pskInfo.cipherSuite0;
+        cs.cipherSuite = pskInfo.cipherSuite;
+
+        if (haveSG && !haveKS) {
```

Comment:
should this be `doKE &&  (!haveSG || !haveKS)`

Reply:
No
https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8
```
   -  If containing a "supported_groups" extension, it MUST also contain
      a "key_share" extension, and vice versa.  An empty
      KeyShare.client_shares vector is permitted.
```